### ISSUE TITLE
Tracking issue for `array::try_map`

### ISSUE THREAD

**Issue by ethanboxx** [rust-lang/rust#79711]

**Comment by arthurlafrance** [comment]
Not sure if this change is still active, but I've stumbled upon a use case where a broader version of this for iterators in general, rather than strictly for arrays, would be useful. Is that something worth considering?

I'm totally new to both Rust contributions & open source in general, so I'm not sure if the idea of a general `try_map` is worth considering/pursuing, but I just thought I'd throw the idea out there since I thought it would be useful. Appreciate any feedback & insights!

(And, if it gets to that point, I'd be interested in taking a stab at implementing it. Would love to make a contribution to Rust, no matter how small!)

**Comment by scottmcm** [comment]
@arthurlafrance It'd suggest starting a conversation on [`IRLO`](https://internals.rust-lang.org/) about it.  Or you can send a PR if it'd be easier to just write the code.

This change is still open, but is for tracking this specific change, not a good place to discuss an additional possible API.

**Comment by marcospb19** [comment]
+1 for this one, it's the only way I found to run a fallible `.map()` without collecting to a `Vec`, and still use `Try` afterward, which leads to a concise and straightforward code.

---

> Not sure if this change is still active, but I've stumbled upon a use case where a broader version of this for iterators in general, rather than strictly for arrays, would be useful.

I guess that's what `.collect::<Result<_>>()` and the unstable `.try_collect()` are intended for.

**Comment by SmnTin** [comment]
Is there anything blocking stabilization?

**Comment by marcospb19** [comment]
@SmnTin so far no blockers and no issues, the _25+_ ":+1:" reactions do indicate "general acceptance", but we need to wait for a team member to start the FCP.

**Comment by scottmcm** [comment]
Note that this, like `array::try_from_fn` and `Iterator::try_find` and similar, depends on the `Residual` trait mechanism for mapping from `Foo<A>` to `Foo<[A; N]>`, which is not used in any stable functions right now, so stabilization here is a bigger question than just the specific function in question.

See https://github.com/rust-lang/rust/pull/94119#issuecomment-1098229331 where `array::from_fn` was accepted but `array::try_from_fn` wasn't accepted due to this.

**Comment by tarcieri** [comment]
I believe #91285 is the tracking issue for stabilizing `Residual`

**Comment by scottmcm** [comment]
Right, stabilizing `try_map` doesn't require stabilizing `Residual` itself, but it does *expose* it, which is itself a commitment to have something like it forever.

**Comment by SmnTin** [comment]
Thank you for the replies! As far as I understood, there is a consensus that exposing `Try` trait in stable APIs is ok (e.g. `Iterator::try_fold` is stable) while exposing `Residual` is not, even though they both belong to `try_trait_v2`.

**Comment by SmnTin** [comment]
In this case, is it possible to rewrite the current implementation to only use `Result`/`Option`, stabilize it and revert back to a more general case when `Residual` trait gets stabilized?

**Comment by eopb** [comment]
I don't think it's possible to rewrite with `Result`/`Option` without changing the number or meaning of `try_map`s generic parameters in a way that won't be backward compatible with the revert.
